server/src/main/java/org/elasticsearch/ingest/IngestService.java

```diff
@@ -875,17 +907,20 @@ private void executePipelines(
                 } catch (IllegalArgumentException ex) {
                     // An IllegalArgumentException can be thrown when an ingest processor creates a source map that is self-referencing.
                     // In that case, we catch and wrap the exception, so we can include more details
-                    listener.onFailure(
-                        new IllegalArgumentException(
-                            format(
-                                "Failed to generate the source document for ingest pipeline [%s] for document [%s/%s]",
-                                pipelineId,
-                                indexRequest.index(),
-                                indexRequest.id()
-                            ),
-                            ex
-                        )
+                    Exception documentContainsSelfReferenceException = new IllegalArgumentException(
+                        format(
+                            "Failed to generate the source document for ingest pipeline [%s] for document [%s/%s]",
+                            pipelineId,
+                            indexRequest.index(),
+                            indexRequest.id()
+                        ),
+                        ex
                     );
+                    if (shouldStoreFailure.test(originalIndex)) {
+                        listener.onResponse(failAndStoreFor(originalIndex, documentContainsSelfReferenceException));
```

Comment:
I don't currently recall where we landed on the pre-transformed versus post-transformed, but storing this document won't try to re-resolve the doc and throw another exception about it being self-referential right? (that would only be the case if we stored post-transformed)

Reply:
We had indeed settled on rolling back ingest changes to the original document state. This is one of those happy outcomes where we don't have to worry about this kind of error causing problems with saving a failure document.